### PR TITLE
Add CI pipeline for Windows Build

### PR DESCRIPTION
**Pull request by toothache**
Add CI pipeline for Windows Build:
- Add script to install vsbuildtools;
- Update `cuda.ps1` to fix VS integration;
- Add CI pipeline for Windows Build.

### CODE REVIEW

File: win/pkgs/cuda.ps1 [hunk 1]
```diff
@@ -25,3 +25,26 @@ Write-Host "Installing CUDA..."
 & "${DownloadDir}/${exe}" -s | Out-Null
 
 ${Env:CUDA_PATH}=[System.Environment]::GetEnvironmentVariable("CUDA_PATH","Machine")
+
+# CUDA MSBuild integration is not working for VS BuildTools.
+# Manually install CUDA Build Customizations files into MSBuild customization folder.
+#
+# See Table 4. CUDA Visual Studio .props locations from
+# https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html
+$vs_where = "${Env:ProgramFiles(x86)}/Microsoft Visual Studio/Installer/vswhere.exe"
```
Comment (xkszltl) [inline win/pkgs/cuda.ps1]: We call this script via setup, so we should already have path here.
Check if it is a build tool installation and if it is already installed.

File: win/pkgs/cuda.ps1 [hunk 2]
```diff
@@ -25,3 +25,26 @@ Write-Host "Installing CUDA..."
...
+if (Test-Path $vs_where)
+{
+    $vs_home = & $vs_where -latest -products 'Microsoft.VisualStudio.Product.BuildTools' -property installationPath
+    if ($vs_home)
+    {
+        Write-Host "Found existing VS installtaion under `"$vs_home`". Skip installation."
```
Comment (xkszltl) [inline win/pkgs/cuda.ps1]: spelling: installation

File: win/pkgs/cuda.ps1 [hunk 3]
```diff
@@ -25,3 +25,26 @@ Write-Host "Installing CUDA..."
...
+
+        $cuda_vsext_dir = "${Env:CUDA_PATH}/extras/visual_studio_integration/MSBuildExtensions"
+        $msbuild_custom_dir = "$vs_home/MSBuild/Microsoft/VC/v160/BuildCustomizations"
```
Comment (xkszltl) [inline win/pkgs/cuda.ps1]: Don't hard code v160 here.

File: win/pkgs/cuda.ps1 [hunk 4]
```diff
@@ -25,3 +25,26 @@ Write-Host "Installing CUDA..."
...
+
+        Write-Host "Patching CUDA VS .props files:"
+        Write-Host "  From: $cuda_vsext_dir"
+        Write-Host "  To:   $msbuild_custom_dir"
+        Copy-Item $cuda_vsext_dir/* $msbuild_custom_dir
```
Comment (xkszltl) [inline win/pkgs/cuda.ps1]: `cp -Force`.
General command preferred over powershell's own name.

File: win/pkgs/vsbuildtools.ps1 [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+{
+    $vs_home = & $vs_where -latest -products * -property installationPath
+    if ($vs_home)
+    {
+        Write-Host "Found existing VS installtaion under `"$vs_home`". Skip installation."
+        exit 0
+    }
+}
+
+Invoke-WebRequest -Uri https://aka.ms/vs/16/release/vs_buildtools.exe -OutFile "${Env:SCRATCH}/vs_buildtools.exe"
```
Comment (xkszltl) [inline win/pkgs/vsbuildtools.ps1]: Quote URL

File: win/pkgs/vsbuildtools.ps1 [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+$install_cmd = "`"${Env:SCRATCH}/vs_buildtools.exe`" --quiet --wait --norestart --nocache " `
```
Comment (xkszltl) [inline win/pkgs/vsbuildtools.ps1]: Simply `cmd /c` here since it is never reused.

Comment (xkszltl) [inline win/pkgs/vsbuildtools.ps1]: 1 args/line.
Don't mix line-breaking styles

Comment (xkszltl) [inline win/pkgs/vsbuildtools.ps1]: sort args if supported.

File: win/pkgs/vsbuildtools.ps1 [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    + "--add Microsoft.VisualStudio.Workload.VCTools;includeRecommended "                   `
+    + "--add Microsoft.VisualStudio.Component.Windows10SDK.16299 "                          `
+    + "--installPath C:\BuildTools || exit /b %ERRORLEVEL%"
+cmd /c $install_cmd
+
+if ($LASTEXITCODE -ne 0 -or $LASTEXITCODE -ne 3003)
```
Comment (xkszltl) [inline win/pkgs/vsbuildtools.ps1]: 1. `$?` instead of `$LASTEXITCODE`
2. Store it in var before use. It is very easy to accidentally change "last".

Comment (xkszltl) [inline win/pkgs/vsbuildtools.ps1]: What is 3003, please explain in comment and link to doc if any.

File: win/pkgs/vsbuildtools.ps1 [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        Write-Host "Found existing VS installtaion under `"$vs_home`". Skip installation."
+        exit 0
+    }
+}
+
+Invoke-WebRequest -Uri https://aka.ms/vs/16/release/vs_buildtools.exe -OutFile "${Env:SCRATCH}/vs_buildtools.exe"
+
+$install_cmd = "`"${Env:SCRATCH}/vs_buildtools.exe`" --quiet --wait --norestart --nocache " `
+    + "--add Microsoft.VisualStudio.Workload.VCTools;includeRecommended "                   `
+    + "--add Microsoft.VisualStudio.Component.Windows10SDK.16299 "                          `
```
Comment (xkszltl) [inline win/pkgs/vsbuildtools.ps1]: Also add latest WinSDK.

File: win/pkgs/vsbuildtools.ps1 [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    + "--installPath C:\BuildTools || exit /b %ERRORLEVEL%"
+cmd /c $install_cmd
```
Comment (xkszltl) [inline win/pkgs/vsbuildtools.ps1]: Why do you need to `cmd /c`?
Should it be `& xxxx | Out-Null`?

File: gitlab-ci/build_stage.ps1 [hunk 10]
```diff
@@ -0,0 +1,39 @@
+Param(
+    [Parameter(Mandatory = $True)]
+    [string] $pkg
+)
+
+$ErrorActionPreference="Stop"
+
+cd $PSScriptRoot/../win
+
+${ENV:SCRATCH} = "${ENV:Agent_TempDirectory}/roaster-scratch"
```
Comment (xkszltl) [inline gitlab-ci/build_stage.ps1]: This dir is specific to gitlab-ci.
I also used it for general purpose script, which is not a good name but doesn't hurt so far.

But for AZP-specific thing, we should definitely create new dir, i.e. `azure-pipelines`.
You can still keep the majority part here, but move things like Agent_TempDirectory there.

Comment (xkszltl) [inline gitlab-ci/build_stage.ps1]: `Env`, not `ENV`, for consistency

File: gitlab-ci/build_stage.ps1 [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+mkdir "${Env:SCRATCH}"
```
Comment (xkszltl) [inline gitlab-ci/build_stage.ps1:14]: Will this work recursively? I mean is it equivalent to `mkdir -p` on Linux?

File: win/pkgs/vsbuildtools.ps1 (hunk 7 above)
Comment (xkszltl) [reply win/pkgs/vsbuildtools.ps1]: `2.` is mostly for multiple uses or `$?` not tightly following is main command.

File: win/pkgs/cuda.ps1 [hunk 12]
```diff
@@ -25,3 +25,26 @@ Write-Host "Installing CUDA..."
 & "${DownloadDir}/${exe}" -s | Out-Null
 
 ${Env:CUDA_PATH}=[System.Environment]::GetEnvironmentVariable("CUDA_PATH","Machine")
+
+# CUDA MSBuild integration is not working for VS BuildTools.
+# Manually install CUDA Build Customizations files into MSBuild customization folder.
+#
+# See Table 4. CUDA Visual Studio .props locations from
```
Comment (xkszltl) [inline win/pkgs/cuda.ps1]: Do not state `4` here.
It is hard to maintain.

Comment (xkszltl) [inline win/pkgs/cuda.ps1]: Add `:` at the end, to make it a more natural line-breaking.

File: azure-pipelines.windows.yml [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+      - template: azure-pipelines/build_stage.windows.yml
+
+    - job: c_ares
+      dependsOn: openssl
+      steps:
+      - template: azure-pipelines/build_stage.windows.yml
+        parameters:
+          stage_name: c-ares
+
+    - job: freetype_fp
```
Comment (xkszltl) [inline azure-pipelines.windows.yml]: No idea what's fp/sp, certainly not first/second pass →_→
Even 1/2 works better...

File: azure-pipelines.windows.yml [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+    - job: mklml
+      dependsOn: intel
+      steps:
+      - template: azure-pipelines/build_stage.windows.yml
+
+- stage: build
+  dependsOn: prep
+  jobs:
+    - job: zlib
```
Comment (xkszltl) [inline azure-pipelines.windows.yml]: Recommend to group pkgs like those in `docker/centos/{infra,util,misc,dl,ort}`, to reduce complexity of `dependsOn` and make it easier to new pkgs.
Your template should be able to handle multiple inputs.

File: azure-pipelines/build_stage.windows.yml [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+  - name: stage_name
+    type: string
+    default: ''
+
+steps:
+- powershell: |
+    if ("${{ parameters.stage_name }}") {
+      .\gitlab-ci\build_stage.ps1 ${{ parameters.stage_name }}
+    } else {
+      .\gitlab-ci\build_stage.ps1 $ENV:SYSTEM_JOBDISPLAYNAME
```
Comment (xkszltl) [inline azure-pipelines/build_stage.windows.yml]: ENV->Env

File: azure-pipelines.windows.yml [hunk 16]
```diff
@@ -0,0 +1,178 @@
+name: Roaster Windows Build
+
+trigger:
+- master
+
+pool:
+  name: $(AGENT_POOL_NAME)
```
Comment (xkszltl) [inline azure-pipelines.windows.yml]: Is this defined by you or by azp?
Default should be public resource, i.e. Microsoft hosted agent (even if there's no enough disk).

Comment (xkszltl) [reply azure-pipelines.windows.yml]: I'm also think about use this in Github action, since the syntax is probably the same.

**Comment by xkszltl** [comment]
@toothache 
Thanks for your contribution!
This is something we want to have since long time ago, but never get the chance to do.

File: win/pkgs/cuda.ps1 (hunk 3 above)
Comment (toothache) [reply win/pkgs/cuda.ps1]: This is required from CUDA documentation:
> Visual Studio 2019 | C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Microsoft\VC\v160\BuildCustomizations

Roaster script is currently hard coding VS2019 as the build tool.

File: win/pkgs/cuda.ps1 (hunk 4 above)
Comment (toothache) [reply win/pkgs/cuda.ps1]: Hmm, I favored powershell own name in a pwsh script. Anyway, I'll follow the conventions here.

File: win/pkgs/vsbuildtools.ps1 (hunk 7 above)
Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: Ah, it should be 3010.

And to be honest, I don't know what 3003 stands for either. I saw it from several places:
- https://docs.microsoft.com/en-us/visualstudio/install/build-tools-container?view=vs-2019
- https://github.com/microsoft/vs-dockerfiles/blob/master/native-desktop/Install.cmd

File: win/pkgs/vsbuildtools.ps1 (hunk 8 above)
Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: It's installed by default with `includeRecommended` option.

You can check the workloads from:
https://docs.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-build-tools?view=vs-2019&preserve-view=true

File: gitlab-ci/build_stage.ps1 (hunk 11 above)
Comment (toothache) [reply gitlab-ci/build_stage.ps1:14]: Yes. mkdir is alias for `New-Item -Type Directory` in pwsh, and it doesn't have a -p switch.

File: azure-pipelines.windows.yml (hunk 13 above)
Comment (toothache) [reply azure-pipelines.windows.yml]: ok.. actually i meant `first_pass` and `second_pass`....

File: azure-pipelines.windows.yml (hunk 16 above)
Comment (toothache) [reply azure-pipelines.windows.yml]: It's a predefined pipeline variable. You should define it in your pipeline.

But azp supports overriding the agent pool. So having this variable here is not required.

File: win/pkgs/cuda.ps1 (hunk 2 above)
Comment (toothache) [reply win/pkgs/cuda.ps1]: good catch

File: win/pkgs/vsbuildtools.ps1 (hunk 7 above)
Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: No, I think we should use $LastExitCode here.

Here're the differences from the [official doc](https://docs.microsoft.com/en-us/previous-versions//dd347675(v=technet.10)?redirectedfrom=MSDN):
- $LastExitCode
   Contains the exit code of the last Windows-based program that was run. 
- $?  
   Contains the execution status of the last operation. It contains TRUE if the last operation succeeded and FALSE if it failed.

From the vsbuildtools installation links I referenced, we should check the error code here.
https://docs.microsoft.com/en-us/visualstudio/install/build-tools-container?view=vs-2019
https://github.com/microsoft/vs-dockerfiles/blob/master/native-desktop/Install.cmd

File: win/pkgs/vsbuildtools.ps1 (hunk 9 above)
Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: hmm, good question.

I had an impression that without calling it in cmd, the `-wait` switch doesn't work. And it looks wierd. I'll double confirm it later.

File: win/pkgs/cuda.ps1 (hunk 4 above)
Comment (toothache) [reply win/pkgs/cuda.ps1]: Fixed.

File: win/pkgs/vsbuildtools.ps1 (hunk 5 above)
Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: Fixed.

File: win/pkgs/vsbuildtools.ps1 (hunk 6 above)
Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: I'm still not quite used to put short args one line each.

Fixed.

Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: Fixed.

File: gitlab-ci/build_stage.ps1 (hunk 10 above)
Comment (toothache) [reply gitlab-ci/build_stage.ps1]: fixed by adding a script param.

Comment (toothache) [reply gitlab-ci/build_stage.ps1]: fixed.

File: win/pkgs/cuda.ps1 (hunk 12 above)
Comment (toothache) [reply win/pkgs/cuda.ps1]: fixed.

File: azure-pipelines.windows.yml (hunk 13 above)
Comment (toothache) [reply azure-pipelines.windows.yml]: fxied.

File: azure-pipelines/build_stage.windows.yml (hunk 15 above)
Comment (toothache) [reply azure-pipelines/build_stage.windows.yml]: fixed.

File: azure-pipelines.windows.yml (hunk 14 above)
Comment (toothache) [reply azure-pipelines.windows.yml]: Good suggestion. Fixed.

File: win/pkgs/vsbuildtools.ps1 (hunk 9 above)
Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: I recalled why i use `cmd/ c` here.

I want to exit with `%ERRORLEVEL%` on non-zero return code.

File: win/pkgs/vsbuildtools.ps1 (hunk 6 above)
Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: Replied in another thread. Resolved.

File: azure-pipelines.windows.yml [hunk 17]
```diff
@@ -0,0 +1,147 @@
+name: Roaster Windows Build
+
+trigger:
+- master
+
+pool:
+  vmImage: 'windows-latest'
```
Comment (toothache) [inline azure-pipelines.windows.yml]: vmImage doesn't work for current selfhost windows agent pool.

should directly use the pool name here.

File: win/pkgs/cuda.ps1 (hunk 3 above)
Comment (xkszltl) [reply win/pkgs/cuda.ps1]: We don't really hard-code it.
We provide a single-place config as env var, so it is more like a config in code.
Adding this kind of assumption in random places with turn that into "hard-code" and make future update much harder.

You should probably list the dir and pick the highest version if multiple.

File: win/pkgs/vsbuildtools.ps1 (hunk 7 above)
Comment (xkszltl) [reply win/pkgs/vsbuildtools.ps1]: Found that in your linked page: 
> Error code 3010 is used to indicate success with a reboot required, see MsiExec.exe error messages for more information.

Please comment in code (in a few words), with the link as well.

Comment (xkszltl) [reply win/pkgs/vsbuildtools.ps1]: And 3010 is probably a failure in our case, since it is still not usable until rebooted.

Comment (toothache) [reply win/pkgs/vsbuildtools.ps1]: Make sense. Remove `3010` error code handling.

File: win/pkgs/cuda.ps1 (hunk 3 above)
Comment (toothache) [reply win/pkgs/cuda.ps1]: Patch every version which has BuildCustomization folder.

File: azure-pipelines.windows.yml [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+      dependsOn: opencv
+      steps:
+      - template: azure-pipelines/build_stage.windows.yml
+
+    - job: onnx
+      dependsOn: rocksdb
+      steps:
+      - template: azure-pipelines/build_stage.windows.yml
+
+    - job: ort
```
Comment (toothache) [inline azure-pipelines.windows.yml:153]: pipeline stuck in `AllocatorTest.CUDAAllocatorTest`, GPU is not available in build agent

Comment (xkszltl) [reply azure-pipelines.windows.yml:153]: Yes, this is a known issue.
I've filed to ort team: https://github.com/microsoft/onnxruntime/issues/4656

Currently I skip those test manually when building CUDA without GPU.

Comment (toothache) [reply azure-pipelines.windows.yml:153]: Seems we don't have a quick fix in ort repo.

Add build option to disable running ort tests in `ort.ps1`.

File: win/pkgs/ort.ps1 [hunk 19]
```diff
@@ -40,6 +40,11 @@ $update_gtest = $true
 $update_onnx = $false
 $update_protobuf = $false
 $use_bat = $false
+$run_tests = $true
+if ($ENV:ROASTER_ORT_SKIP_TESTS)
```
Comment (xkszltl) [inline win/pkgs/ort.ps1]: 1. ${Env:ROASTER_ORT_SKIP_TESTS}
2. There's no point to use both env var and var.
3. Please use auto detection instead. i.e. with nvidia-smi and nvidia-smi -L

**Comment by xkszltl** [comment]
@toothache 
I saw you have a fix (https://github.com/microsoft/onnxruntime/pull/6138) for ort cuda unittest.
You can open a new PR to cherry-pick that into both windows and linux build.

File: win/pkgs/ort.ps1 (hunk 19 above)
Comment (toothache) [reply win/pkgs/ort.ps1]: Fixed in latest update.

The option is detected using:
`$((Get-Command -Name "nvidia-smi" -ErrorAction SilentlyContinue) -and (nvidia-smi -L))`

**Comment by toothache** [comment]
> @toothache
> I saw you have a fix ([microsoft/onnxruntime#6138](https://github.com/microsoft/onnxruntime/pull/6138)) for ort cuda unittest.
> You can open a new PR to cherry-pick that into both windows and linux build.

I'll do that once that PR is merged.

**Pull request merged**